lib/rubocop/cop/rspec/leaky_local_variable.rb

```diff
@@ -0,0 +1,138 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module RSpec
+      # Checks for outside local variables used in examples.
+      #
+      # Local variables assigned outside an example but used within it act
+      # as shared state, which can make tests non-deterministic.
+      #
+      # @example
+      #   # bad - outside variable used in a hook
+      #   user = create(:user)
+      #
+      #   before { user.update(admin: true) }
+      #
+      #   # good
+      #   let(:user) { create(:user) }
+      #
+      #   before { user.update(admin: true) }
+      #
+      #   # bad - outside variable used in an example
+      #   user = create(:user)
+      #
+      #   it 'is persisted' do
+      #     expect(user).to be_persisted
+      #   end
+      #
+      #   # good
+      #   let(:user) { create(:user) }
+      #
+      #   it 'is persisted' do
+      #     expect(user).to be_persisted
+      #   end
+      #
+      #   # also good - assigning the variable within the example
+      #   it 'is persisted' do
+      #     user = create(:user)
+      #
+      #     expect(user).to be_persisted
+      #   end
+      #
+      #   # bad - outside variable passed to included examples
+      #   attrs = ['foo', 'bar']
+      #
+      #   it_behaves_like 'some examples', attrs
+      #
+      #   # good
+      #   it_behaves_like 'some examples' do
+      #     let(:attrs) { ['foo', 'bar'] }
+      #   end
+      #
+      #   # good - when variable is used only as example description
+      #   attribute = 'foo'
+      #
+      #   it "#{attribute} is persisted" do
+      #     expectations
+      #   end
+      #
+      #   # good - when variable is used only to include other examples
+      #   examples = foo ? 'some examples' : 'other examples'
+      #
+      #   it_behaves_like examples, another_argument
+      #
+      class LeakyLocalVariable < Base
+        MSG = 'Do not use local variables defined outside of ' \
+              'examples inside of them.'
+
+        # @!method example_method?(node)
+        def_node_matcher :example_method?, <<~PATTERN
+          (send nil? #Examples.all _)
+        PATTERN
+
+        # @!method includes_method?(node)
+        def_node_matcher :includes_method?, <<~PATTERN
+          (send nil? #Includes.all ...)
+        PATTERN
+
+        def self.joining_forces
+          VariableForce
+        end
+
+        def after_leaving_scope(scope, _variable_table)
+          scope.variables.each_value { |variable| check_references(variable) }
+        end
+
+        private
+
+        def check_references(variable)
+          variable.assignments.each do |assignment|
+            next if part_of_example_scope?(assignment.node)
```

Comment:
Can those be computed once per `scope`? Is this a worthwhile optimization? It feels like we’re iterating over vars defined in the same scope, right?